### PR TITLE
PIO2 Micro timers

### LINKED ISSUE #11: Adding micro timers in PIO2
Add micro timers in PIO1 (in acme_develop_pio1_0 branch) to PIO2.
This enhancement prints out read and write timing statistics from 
io rank 0 to a pio log file (piowinfo*.dat).
The timings here are more fine grained than what we capture in
ACME performance logs.

### PR DESCRIPTION
**Pull request by jayeshkrishna**
Adding micro timers. These timers give timing information about
read and write times per variable (for each read/write) and the
data rearrange times.

To enable micro timers users need to add 
"-DPIO_MICRO_TIMING:BOOL=ON" when configuring PIO2. 
The statistics are written out to "piorwinfo*.dat" .

Fixes #11

### CODE REVIEW

**Comment by jayeshkrishna** [comment]
The PIO2 tests were run successfully on laptop, compute001 (anlworkstation) and anvil.

File: src/clib/pio_darray.c [hunk 1]
```diff
@@ -222,6 +228,38 @@ int PIOc_write_darray_multi(int ncid, const int *varids, int ioid, int nvars,
      * method.  */
     rlen = iodesc->maxiobuflen * nvars;
 
+#ifdef PIO_MICRO_TIMING
+    bool var_mtimer_was_running[nvars];
```
Comment (dqwu) [inline src/clib/pio_darray.c]: This definition can be moved to the body of the if statement below.

File: src/clib/pio.h [hunk 2]
```diff
@@ -107,6 +245,9 @@ typedef struct var_desc_t
 {
     /* Variable ID. */
     int varid;
+
+    /* Variable name - cached */
+    char vname[PIO_MAX_NAME];
```
Comment (dqwu) [inline src/clib/pio.h]: Should we use [PIO_MAX_NAME + 1] here? In line 760 of pioc_nc.c "char my_name[PIO_MAX_NAME + 1];" is used.

File: src/clib/pio_nc.c [hunk 3]
```diff
@@ -812,10 +820,40 @@ int PIOc_inq_var(int ncid, int varid, char *name, nc_type *xtypep, int *ndimsp,
...
+                {
+                    return pio_err(ios, file, PIO_ENOMEM, __FILE__, __LINE__);
+                }
+            }
             if (!ierr)
-                ierr = ncmpi_inq_var(file->fh, varid, name, xtypep, ndimsp, dimidsp, nattsp);
+                ierr = ncmpi_inq_var(file->fh, varid, my_name, xtypep, ndimsp, (dimidsp)?dimidsp:tmp_dimidsp, nattsp);
+            if (!ierr && name)
+            {
+                strncpy(name, my_name, PIO_MAX_NAME);
```
Comment (dqwu) [inline src/clib/pio_nc.c]: Is there a reason that we should read var name to my_name first  andthen strncpy to name? The original code directly reads var name to name without the overhead of strncpy.

File: src/clib/pio_timer.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+} mtimer_type_t;
+
+/* FIXME: Do we need to return error codes here - or just have asserts */
+/* Init timer framework - needs to be called once before using timers */
+int mtimer_init(mtimer_type_t type);
+/* Create/Start/Stop/Destroy a timer */
+mtimer_t mtimer_create(const char *name, MPI_Comm comm, char *log_fname);
+int mtimer_start(mtimer_t mt);
+int mtimer_stop(mtimer_t mt, const char *log_msg);
+int mtimer_destroy(mtimer_t *mt);
```
[the remark this answers is not in the document]
Comment (jayeshkrishna) [reply src/clib/pio_timer.h]: ok, will do

File: src/clib/pio_darray.c (hunk 1 above)
Comment (jayeshkrishna) [reply src/clib/pio_darray.c]: ok, will do

File: src/clib/pio.h (hunk 2 above)
Comment (jayeshkrishna) [reply src/clib/pio.h]: I somehow don't like the PIO_MAX_NAME+1 style (add 1 for the string delimiter). It shouldn't be an issue unless we have a very long variable name and strncpy() does not add the delimiter. In any case we won't leak.

File: src/clib/pio_nc.c (hunk 3 above)
Comment (jayeshkrishna) [reply src/clib/pio_nc.c]: There could be issues when user does not want to inquire the variable name (name is NULL)

File: src/clib/pio_darray.c (hunk 1 above)
Comment (jayeshkrishna) [reply src/clib/pio_darray.c]: Not sure if this can be done, var_mtimer_was_running is required later (line 357)

File: src/clib/pio_nc.c (hunk 3 above)
Comment (dqwu) [reply src/clib/pio_nc.c]: When name is NULL, name_present will be set to false and broadcast
We can use the logic below
if (!ierr && name) ierr = ncmpi_inq_var(file->fh, varid, name, ...)
This way we still do not need my_name (which will always be read by ncmpi_inq_var)

File: src/clib/pio_darray.c (hunk 1 above)
Comment (dqwu) [reply src/clib/pio_darray.c]: Just noticed this code at the beginning of PIOc_write_darray_multi:
    if (nvars <= 0 || !varids)
        return pio_err(ios, file, PIO_EINVAL, __FILE__, __LINE__);
There you do not need to test "if (nvars > 0)" below.

File: src/clib/pio_nc.c (hunk 3 above)
Comment (jayeshkrishna) [reply src/clib/pio_nc.c]: We always (independent of user choice) need to query the name (to cache it)

File: src/clib/pio_darray.c (hunk 1 above)
Comment (jayeshkrishna) [reply src/clib/pio_darray.c]: ok, will remove the extra check (nvars > 0)

**Comment by jayeshkrishna** [comment]
@dqwu : Please take a look at the changes (force pushed the branch with changes).

**Comment by jayeshkrishna** [comment]
The PIO2 tests run successfully on Anvil after the changes.

**Comment by jayeshkrishna** [comment]
(I should have waited for nightlies to run before merging to master. In any case, now I have merged the branch to develop and should show up in the next nightly runs)

**Pull request merged**